src/Universum/String/Conversion.hs

```diff
@@ -164,41 +164,49 @@ instance ToString T.Text where
 instance ToString LT.Text where
     toString = LT.unpack
 
+{- [Note toString-toText-rewritting]
+
+Note ON MAINTENANCE of rewrite rules below.
+
+Whenever you want to allow a newer version of @text@ package, check:
+* The @text conversions@ benchmark, that the numbers with and without the rules
+are still the expected ones (see the comments there for what to expect).
+* You may optionally check whether any changes to `pack` and `unpack` functions
+and their `INLINE`/`NOINLINE` annotations took place.
+The current rewrite rules match with what happens in `text` package
+at the e5644b663c32c01a1de7299a5e711216755e01bc commit, and next time
+you can just check `git diff <that commit>..HEAD`.
+
+If these points hold, it should be safe to raise the upper bound on `text` version.
+
+If not, first do the necessary changes (preserving the backward compatibility),
+update the commit id above if you checked the code of the `text` package.
+Then bump the version constraint.
+-}
+
 {-
 
 @toString . toText@ pattern may occur quite often after inlining because
 we tend to use 'Text' rather than 'String' in function signatures, but
 there are still some libraries which use 'String's and thus make us perform
 conversions back and forth.
 
-Note that @toString . toText@ is not strictly equal to identity function, see
-explanation in the comment below.
+Note that, as documentation for 'T.pack' function mentions, @toString . toText@
+is not strictly equal to the identity function. Thus, replacing @toString . toText@
+with @id@ will result in losing transformation of surrogate code points.
+But in the most cases this is what the user wants.
 -}
 
 {-# RULES "pack/unpack" [~0]
     forall s. T.unpack (T.pack s) = s
 #-}
 
-{- [Note toString-toText-rewritting]
+#if !MIN_VERSION_text(2,0,2)
```

Comment:
I don't exactly remember the behaviour of `MIN_VERSION_*` macros, but AFAICT this means "when text < 2.0.2". This looks extremely suspicious to me.

Reply:
Yes, this is what I wanted. The rewrite rules below don't work starting from 2.0.2, while the remaining rules should work much better now (previously I had those "just in case").

Reply:
That's a true :ssc:, I really wished I could use some `MAX_VERSION` instead (but it seems to not exist)